cdispyutils/auth/jwt_validation.py

```diff
@@ -0,0 +1,234 @@
+    """
+    aud = set(aud)
+    if not aud:
+        raise JWTAudienceError('no audiences provided')
+    request = request or flask.request
+    try:
+        encoded_token = request.headers['Authorization'].split(' ')[1]
+    except KeyError:
+        raise JWTValidationError('no authorization token provided')
+    token_headers = jwt.get_unverified_header(encoded_token)
+    public_key = get_public_key_for_kid(token_headers.get('kid'))
+    return validate_jwt(encoded_token, public_key, aud)
+
+
+def require_jwt(aud):
+    """
+    Define a decorator for utility which calls ``validate_request_jwt`` using
+    the given audiences to require a JWT authorization header.
+
+    Args:
+        aud (Iterable[str]): audiences to require in the JWT
+
+    Return:
+        Callable[[Any], Any]: decorated function
+    """
+    def decorator(f):
+        """
+        Define the actual decorator, which takes the decorated function as an
+        argument.
+
+        Args:
+            Callable[[Any], Any]: function to decorate
+
+        Return:
+            Callable[[Any], Any]: the same type as the function
+        """
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            """Validate the JWT and call the actual function here."""
+            validate_request_jwt(aud)
+            return f(*args, **kwargs)
+        return wrapper
+    return decorator
+
+
+def validate_jwt(encoded_token, public_key, aud):
+    """
+    Validate the encoded JWT ``encoded_token``, which must satisfy the
+    audiences ``aud``.
+
+    - Decode JWT using public key; PyJWT will fail if iat or exp fields are
+      invalid
+    - Check audiences: token audiences must be a superset of required audiences
+      (the ``aud`` argument); fail if not satisfied
+
+    Args:
+        encoded_token (str): encoded JWT
+        public_key (str): public key to validate the JWT signature
+        aud (set): non-empty set of audiences the JWT must satisfy
+
+    Return:
+        dict: the decoded and validated JWT
+
+    Raises:
+        JWTValidationError: if any step of the validation fails
+    """
+    # To satisfy PyJWT, since the token will contain an aud field, decode has
+    # to be passed one of the audiences to check here (so PyJWT doesn't raise
+    # an InvalidAudienceError). Per the JWT specification, if the token
+    # contains an aud field, the validator MUST identify with one of the
+    # audiences listed in that field. This implementation is more strict, and
+    # allows the validator to demand multiple audiences which must all be
+    # satisfied by the token (see below).
+    aud = set(aud)
+    random_aud = list(aud)[0]
+    try:
+        token = jwt.decode(
+            encoded_token, key=public_key, algorithms=['RS256'],
+            audience=random_aud
+        )
+    except jwt.InvalidAudienceError as e:
+        raise JWTAudienceError(e)
+
+    # PyJWT validates iat and exp fields (and aud...sort of); everything else
```

Comment:
verify the issuer is the config[USER_API]?

Reply:
added in 61f0957

Reply:
added test in f466758